include_physics/ideal_MHD.h

```diff
@@ -388,13 +388,180 @@ inline void init_fields(uint ix, uint iy, uint iz, global REAL* u) {
+  else if(0 < ar) {
+    for (i = 0; i < NUM_CONSERVED_VARS; i++)
+      num_flux[i] = ((ar*fluxl[i]-al*fluxr[i]) + ar*al*(ur[i]-ul[i]))/(ar-al);
+  } 
+  else {
+    for (i = 0; i < NUM_CONSERVED_VARS; i++)
+      num_flux[i] = fluxr[i];
+  }
+}
+
+#endif
+
 inline void add_surface_terms(REAL time, uint ix, uint iy, uint iz, global REAL *u, REAL *du_dt) {
 
   // For periodic boundary conditions and a single block, no surface term has to be used.
 #ifndef USE_PERIODIC
-  #error "Error in ideal_MHD.cl: Surface terms not implemented for nonperiodic boundaries!"
-#endif // USE_PERIODIC*/
+  int i;
+  REAL um[NUM_TOTAL_VARS] = {0.0};
+  get_field(ix, iy, iz, 0, 0, 0, u, um); 
+  REAL4 b_bound = b_boundary(ix, iy, iz, time);
+  REAL4 u_bound = u_boundary(ix, iy, iz, time);
+  REAL rho_bound = (REAL)1;
+  REAL p_bound = 1;
+  REAL E_bound = compute_energy(p_bound, rho_bound, u_bound.x, u_bound.y, u_bound.z, b_bound.x, b_bound.y, b_bound.z);
+
+
+  REAL ub[NUM_TOTAL_VARS] = {0.0};
+  ub[Field_ux] = u_bound.x;
+  ub[Field_uy] = u_bound.y;
+  ub[Field_uz] = u_bound.z;
+  ub[Field_rho_ux] = u_bound.x * rho_bound;
+  ub[Field_rho_uy] = u_bound.y * rho_bound;
+  ub[Field_rho_uz] = u_bound.z * rho_bound;
+  ub[Field_rho] = rho_bound;
+  ub[Field_p] = p_bound;
+  ub[Field_E] = E_bound;
+  ub[Field_Bx] = b_bound.x;
+  ub[Field_By] = b_bound.y;
+  ub[Field_Bz] = b_bound.z;
+
+  REAL flux[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxm[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxb[NUM_CONSERVED_VARS] = {0.0};
+
+  // Speeds for the HLL solver are from equation (12) of 
+  // A multi-state HLL approximate Riemann solver for ideal magnetohydrodynamics
+  // Miyoshi and Kusano (2005)
+  REAL BBSQb = ub[Field_Bx] * ub[Field_Bx] + ub[Field_By] * ub[Field_By] + ub[Field_Bz] * ub[Field_Bz];
+  REAL BBSQm = um[Field_Bx] * um[Field_Bx] + um[Field_By] * um[Field_By] + um[Field_Bz] * um[Field_Bz];
+
+
+  // X direction
+  REAL cfbx = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_Bx] * ub[Field_Bx]))/(2 * ub[Field_rho]));
+
+  REAL cfmx = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_Bx] * um[Field_Bx]))/(2 * um[Field_rho]));
+
+  // Y direction
+  REAL cfby = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_By] * ub[Field_By]))/(2 * ub[Field_rho]));
+
+  REAL cfmy = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_By] * um[Field_By]))/(2 * um[Field_rho]));
+
+  // Z direction
+  REAL cfbz = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(pow(GAMMA * ub[Field_p] + BBSQb, 2) - 4 * GAMMA * ub[Field_p] * ub[Field_Bz] * ub[Field_Bz]))/(2 * ub[Field_rho]));
+
+  REAL cfmz = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(pow(GAMMA * um[Field_p] + BBSQm, 2) - 4 * GAMMA * um[Field_p] * um[Field_Bz] * um[Field_Bz]))/(2 * um[Field_rho]));
```

Comment:
These calls to `pow,(..., 2)` should also be replaced.

Reply:
I used as Macro to retain the legibility of the Code, i hope this is okay.